### PR TITLE
Feat/file api

### PR DESCRIPTION
**Pull request by Kerwin-Jhong**
Fixes #

### CODE REVIEW

File: .projenrc.js [hunk 1]
```diff
@@ -24,8 +24,10 @@ const project = new AwsCdkConstructLibrary({
     'joi',
     'semver',
     'uuid',
+    'lodash',
```
Comment (Kerwin-Jhong) [inline .projenrc.js:27]: 確認是否有用到

File: .projenrc.js [hunk 2]
```diff
@@ -24,8 +24,10 @@ const project = new AwsCdkConstructLibrary({
...
     '@types/node@15.12.2',
     '@types/uuid',
+    '@types/lodash',
```
Comment (Kerwin-Jhong) [inline .projenrc.js:30]: 確認是否有用到

File: .projenrc.js (hunk 1 above)
Comment (liaojason2softchef) [reply .projenrc.js]: 沒用到

File: .projenrc.js (hunk 2 above)
Comment (liaojason2softchef) [reply .projenrc.js]: 沒用到

File: lambda-assets/files/create-file/app.ts [hunk 3]
```diff
@@ -1,52 +1,108 @@
...
+          joi.object().keys({
+            location: joi.string().uri().required(),
+            checksumType: joi.string().allow('md5', 'crc32', 'sha1'),
+            checksum: joi.string().when('checksumType', { is: 'md5', then: joi.string().length(32).required() })
+              .concat(joi.string().when('checksumType', { is: 'crc32', then: joi.string().length(8).required() }))
+              .concat(joi.string().when('checksumType', { is: 'sha1', then: joi.string().length(40).required() })),
+            version: joi.string().required(),
+            categoryId: joi.string().required(),
+            locales: joi.array().items(
+              joi.object().keys({
```
Comment (MinCheTsai) [inline lambda-assets/files/create-file/app.ts:22]: locale is string, not array object

File: lambda-assets/files/delete-category/app.ts [hunk 4]
```diff
@@ -22,5 +22,5 @@ export async function handler(event: { [key: string]: any }) {
     });
   } catch (error) {
     return response.error(error);
-  }
+  };
```
Comment (MinCheTsai) [inline lambda-assets/files/delete-category/app.ts:25]: lint

File: src/file-api.ts [hunk 5]
```diff
@@ -109,17 +125,17 @@ export class FileApi extends cdk.Construct {
           lambdaFunction: this.createListFilesFunction(),
         },
         {
-          path: '/files/{fileId}/versions/{version}',
+          path: '/files/{checksum}/versions/{version}',
```
Comment (MinCheTsai) [inline src/file-api.ts:128]: fileId為何要改掉？

File: src/file-api.ts [hunk 6]
```diff
@@ -46,23 +46,39 @@ export class FileApi extends cdk.Construct {
         name: 'fileId',
         type: dynamodb.AttributeType.STRING,
       },
+      readCapacity: 1,
```
Comment (MinCheTsai) [inline src/file-api.ts:49]: 抽成外部參數
```
new FileApi(this, 'FileAPI', {
  {index}?: {
      WriteCapacityUnit?: 1,
      ReadCapacityUnit?: 1
  }
})
```

File: docs/file-api.md [hunk 7]
```diff
@@ -46,7 +46,7 @@ Create new file category
 | ---------------- | ----------- |
 | 200 | Create success |
 | 404 | Parent category id not found |
-| 422 | Missing require field / Variable type incorrect |
+| 422 | Category already exists / Variable type incorrect |
```
Comment (MinCheTsai) [inline docs/file-api.md:49]: Missing require field也要加上

File: docs/file-api.md [hunk 8]
```diff
@@ -88,7 +88,7 @@ Create files
 | HTTP Status Code | Description |
 | ---------------- | ----------- |
 | 200 | Create file success |
-| 422 | Missing require field / Variable type incorrect / File exists |
+| 422 | Variable type incorrect |
```
Comment (MinCheTsai) [inline docs/file-api.md:91]: 依照status code順序

File: docs/file-api.md [hunk 9]
```diff
@@ -402,32 +402,24 @@ Update category description by category ID
 | ---------------- | ----------- |
 | 200 | Update success |
 | 404 | Category ID not found |
-| 422 | Missing require field / Variable type incorrect |
+| 422 | Variable type incorrect |
```
Comment (MinCheTsai) [inline docs/file-api.md:405]: 沒有必填欄位嗎？

File: lambda-assets/files/list-files-by-category/app.ts [hunk 10]
```diff
@@ -20,13 +20,15 @@ export async function handler(event: { [key: string]: any }) {
...
-        KeyConditionExpression: '#categoryId = :categoryId',
+        IndexName: 'query-by-category-id-and-locale',
+        KeyConditionExpression: '#categoryId = :categoryId and #locale = :locale',
         ExpressionAttributeNames: {
           '#categoryId': 'categoryId',
+          '#locale': 'locale',
         },
         ExpressionAttributeValues: {
           ':categoryId': request.parameter('categoryId'),
+          ':locale': request.get('locale'),
```
Comment (MinCheTsai) [inline lambda-assets/files/list-files-by-category/app.ts:31]: locale是選填，非必要

File: lambda-assets/files/list-files-by-category/app.ts [hunk 11]
```diff
@@ -36,7 +38,7 @@ export async function handler(event: { [key: string]: any }) {
       nextToken = Buffer.from(
         JSON.stringify(lastEvaluatedKey),
       ).toString('base64');
-    }
+    };
```
Comment (MinCheTsai) [inline lambda-assets/files/list-files-by-category/app.ts:41]: lint

File: lambda-assets/files/list-files/app.ts [hunk 12]
```diff
@@ -28,7 +28,7 @@ export async function handler(event: { [key: string]: any }) {
       nextToken = Buffer.from(
         JSON.stringify(lastEvaluatedKey),
       ).toString('base64');
-    }
+    };
```
Comment (MinCheTsai) [inline lambda-assets/files/list-files/app.ts:31]: lint

File: lambda-assets/files/update-category/app.ts [hunk 13]
```diff
@@ -17,29 +17,30 @@ export async function handler(event: { [key: string]: any }) {
...
+          categoryId,
         },
       }),
     );
-    if (!category) {
-      return response.error('Not found.', 404);
-    }
+    if (!category.Item) {
+      return response.error('Category does not exist.', 404);
+    };
```
Comment (MinCheTsai) [inline lambda-assets/files/update-category/app.ts:31]: lint

File: src/file-api.ts [hunk 14]
```diff
@@ -109,17 +135,22 @@ export class FileApi extends cdk.Construct {
...
         },
         {
-          path: '/files/{fileId}/versions/{version}',
+          path: '/files/{fileId}',
           httpMethod: HttpMethod.GET,
           lambdaFunction: this.createGetFileFunction(),
         },
         {
-          path: '/files/{fileId}/versions/{version}',
+          path: '/files/{checksum}/versions/{version}',
```
Comment (MinCheTsai) [inline src/file-api.ts:143]: 順序依照路徑排序

File: src/file-api.ts [hunk 15]
```diff
@@ -109,17 +135,22 @@ export class FileApi extends cdk.Construct {
...
+          lambdaFunction: this.createGetFilesFunction(),
+        },
+        {
+          path: '/files/{fileId}',
           httpMethod: HttpMethod.PUT,
           lambdaFunction: this.createUpdateFileFunction(),
         },
         {
-          path: '/files/{fileId}/versions/{version}',
+          path: '/files/{checksum}/versions/{version}',
```
Comment (MinCheTsai) [inline src/file-api.ts:153]: 最終是？

File: lambda-assets/files/list-categories/app.ts [hunk 16]
```diff
@@ -12,29 +20,27 @@ export async function handler(event: { [key: string]: any }) {
...
-    }
     if (request.has('parentId')) {
-      parameters.IndexName = 'query-by-parent-id';
-      parameters.KeyConditionExpression = '#parentId = :parentId';
-      parameters.ExpressionAttributeNames = {
-        '#parentId': 'parentId',
-      };
-      parameters.ExpressionAttributeValues = {
-        ':parentId': parentId,
+      let parameters: QueryCommandInput = {
```
Comment (MinCheTsai) [inline lambda-assets/files/list-categories/app.ts:24]: const

File: lambda-assets/files/create-file/app.ts [hunk 17]
```diff
@@ -27,31 +35,33 @@ export async function handler(event: { [key: string]: any }) {
...
+      TableName: process.env.CATEGORY_TABLE_NAME,
+      Key: {
+        categoryId,
+      },
+    };
+    const category = await ddbDocClient.send(
+      new GetCommand(getParameters),
     );
-    if (!category) {
+    if (!category || !category.Item) {
```
Comment (MinCheTsai) [inline lambda-assets/files/create-file/app.ts:47]: { Item: category}

File: lambda-assets/files/get-file/app.ts [hunk 18]
```diff
@@ -12,16 +18,17 @@ export async function handler(event: { [key: string]: any }) {
...
+      Key: {
+        fileId: request.parameter('fileId'),
+        version: request.parameter('version'),
+      },
+    };
+    const file = await ddbDocClient.send(
+      new GetCommand(getParameters),
     );
-    if (!file) {
+    if (!file || !file.Item) {
```
Comment (MinCheTsai) [inline lambda-assets/files/get-file/app.ts:31]: { Item: file }

File: lambda-assets/files/update-category/app.ts [hunk 19]
```diff
@@ -20,31 +28,34 @@ export async function handler(event: { [key: string]: any }) {
...
-        },
-      }),
+    const categoryId = request.parameter('categoryId');
+    const getParameters: GetCommandInput = {
+      TableName: process.env.CATEGORY_TABLE_NAME,
+      Key: {
+        categoryId,
+      },
+    };
+    const category = await ddbDocClient.send(
```
Comment (MinCheTsai) [inline lambda-assets/files/update-category/app.ts:38]: { Item: category }

File: src/file-api.ts [hunk 20]
```diff
@@ -19,6 +19,16 @@ export interface FileApiProps {
    * @default undefined
    */
   readonly authorizer?: apigateway.IAuthorizer | undefined;
+  readonly categoryTableRCU?: number | 1;
```
Comment (MinCheTsai) [inline src/file-api.ts:22]: interface CategoryTableConfig {
  readonly primaryIndex: {
    writeCapacity: Number;
    readCapacity: Number;
  };
  readonly queryByParentId: {
    writeCapacity: Number;
    readCapacity: Number;
  };
}

interface FileApiProps {
  readonly categoryTable: CategoryTableConfig;
}

File: src/file-api.ts [hunk 21]
```diff
@@ -386,22 +432,22 @@ export class FileApi extends cdk.Construct {
...
-      entry: `${LAMBDA_ASSETS_PATH}/delete-file/app.ts`,
+  private createDeleteFilesFunction(): lambda.NodejsFunction {
+    const deletesFileFunction = new lambda.NodejsFunction(this, 'DeletesFileFunction', {
+      entry: `${LAMBDA_ASSETS_PATH}/delete-files/app.ts`,
       environment: {
         FILE_TABLE_NAME: this.fileTable.tableName,
       },
     });
-    deleteFileFunction.role?.attachInlinePolicy(
+    deletesFileFunction.role?.attachInlinePolicy(
```
Comment (MinCheTsai) [inline src/file-api.ts:445]: deleteFilesFunction

**Pull request merged**